Add workflow using conda to install dependencies

diegoferigo opened the pull request:
#279 provides useful information about why relying on conda could be beneficial over the current PyPI-based distribution.

Much work has been done by the conda-forge community (including @traversaro) to provide the Ignition stack:

- https://github.com/conda-forge/libignition-physics-feedstock
- https://github.com/conda-forge/libignition-gazebo-feedstock
- https://github.com/conda-forge/libignition-rendering4-feedstock/
- ... and much more

This PR explores the integration of conda and its channels (particularly, conda-forge and robotology) to install the build dependencies in CI.

- Only Dome is supported (Edifice has not been released yet, and it's not yet part of conda-forge)
- The iDynTree dependency in conda channels does not have yet the python component https://github.com/robotology/robotology-superbuild/issues/641, as temporarily workaround it gets installed with pip from the git repository
- For some reason, exporting `CMAKE_PREFIX_PATH` for iDynTree does not work in this setup, I had to rely on `iDynTree_DIR` (note that iDynTree exported targets are inside the python package)
- Only Ubuntu Focal is supported even though there are already unused steps for macOS and Windows

This PR is currently blocked by https://github.com/conda-forge/libignition-gazebo-feedstock/issues/4.

traversaro commented:
> The iDynTree dependency in conda-forge does not have yet the python component robotology/robotology-superbuild#641, as temporarily workaround it gets installed with pip from the git repository

In general FYI I am all in favor in progressively moving dependencies with a clear versioning API/ABI and release policy from the robotology channel to proper conda-forge dependencies. iDynTree is definitely one of the first candidates, the missing part is just to move the idyntree-yarp-tools to https://github.com/robotology/idyntree-yarp-tools .

diegoferigo commented:
> > The iDynTree dependency in conda-forge does not have yet the python component [robotology/robotology-superbuild#641](https://github.com/robotology/robotology-superbuild/issues/641), as temporarily workaround it gets installed with pip from the git repository
> 
> In general FYI I am all in favor in progressively moving dependencies with a clear versioning API/ABI and release policy from the robotology channel to proper conda-forge dependencies. iDynTree is definitely one of the first candidates, the missing part is just to move the idyntree-yarp-tools to https://github.com/robotology/idyntree-yarp-tools .

My comment was not clear enough: substitute `conda-forge` -> `robotology` channel. What I meant is that also in the `robotology` channel the Python components are missing. For the sake of this PR, there's no need to move anything to `conda-forge`.

traversaro commented:
> 
> 
> > > The iDynTree dependency in conda-forge does not have yet the python component [robotology/robotology-superbuild#641](https://github.com/robotology/robotology-superbuild/issues/641), as temporarily workaround it gets installed with pip from the git repository
> > 
> > 
> > In general FYI I am all in favor in progressively moving dependencies with a clear versioning API/ABI and release policy from the robotology channel to proper conda-forge dependencies. iDynTree is definitely one of the first candidates, the missing part is just to move the idyntree-yarp-tools to https://github.com/robotology/idyntree-yarp-tools .
> 
> My comment was not clear enough: substitute `conda-forge` -> `robotology` channel. What I meant is that also in the `robotology` channel the Python components are missing. For the sake of this PR, there's no need to move anything to `conda-forge`.

Actually I did not even noticed the conda-forge/robotology typo. I just mentioned that for mutual alignment.

diegoferigo commented:
I re-triggered CI just to check and now after, the error of the wrong plugin directory seems solved. The project is building (also before it was) and the tests run and fail. This time, the error is the following:

> Error while loading the library [/usr/share/miniconda/envs/ci/lib/ign-physics-4/engine-plugins/libignition-physics-dartsim-plugin.so]: libode.so.0.16.2: cannot open shared object file: No such file or directory
> [Err] [Physics.cc:545] Unable to load the [/usr/share/miniconda/envs/ci/lib/ign-physics-4/engine-plugins/libignition-physics-dartsim-plugin.so] library.

Inspecting the conda environment, no ODE package is installed automatically (this was what I was mentioning today about the runtime dependencies of ignition libraries packaged in conda-forge @traversaro).

<details>
<summary>mamba list</summary>

```
                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.15.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████

# packages in environment at /usr/share/miniconda/envs/ci:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
_sysroot_linux-64_curr_repodata_hack 3                   h5bd9786_12    conda-forge
alsa-lib                  1.2.3                h516909a_0    conda-forge
ampl-mp                   3.1.0             h616b090_1004    conda-forge
assimp                    5.0.1                hedfc422_5    conda-forge
attrs                     21.2.0             pyhd8ed1ab_0    conda-forge
binutils                  2.36.1               hdd6e379_1    conda-forge
binutils_impl_linux-64    2.36.1               h193b22a_1    conda-forge
binutils_linux-64         2.36                hf3e587d_32    conda-forge
boost-cpp                 1.74.0               h312852a_4    conda-forge
bullet-cpp                3.17                 h1abd341_0    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.17.1               h7f98852_1    conda-forge
c-compiler                1.2.0                h7f98852_0    conda-forge
ca-certificates           2021.5.30            ha878542_0    conda-forge
certifi                   2021.5.30        py38h578d9bd_0    conda-forge
cli11                     2.0.0                h9c3ff4c_0    conda-forge
cloudpickle               1.6.0                      py_0    conda-forge
python_abi                3.8                      2_cp38    conda-forge
qt                        5.12.9               hda022c4_4    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
rhash                     1.4.1                h7f98852_0    conda-forge
ruby                      2.7.2                he592edb_3    conda-forge
scipy                     1.7.0            py38h7b17777_1    conda-forge
scotch                    6.0.9                h3858553_1    conda-forge
sdl2                      2.0.12               h9c3ff4c_1    conda-forge
setuptools                49.6.0           py38h578d9bd_3    conda-forge
sqlite                    3.36.0               h9cd32fc_0    conda-forge
swig                      4.0.2                hd3c618e_2    conda-forge
sysroot_linux-64          2.17                h4a8ded7_12    conda-forge
tinyxml                   2.6.2                h4bd325d_2    conda-forge
tinyxml2                  9.0.0                h9c3ff4c_1    conda-forge
tk                        8.6.10               h21135ba_1    conda-forge
toml                      0.10.2             pyhd8ed1ab_0    conda-forge
urdfdom                   2.3.5                h4bd325d_0    conda-forge
urdfdom_headers           1.0.5                h4bd325d_2    conda-forge
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
x264                      1!161.3030           h7f98852_1    conda-forge
xorg-kbproto              1.0.7             h7f98852_1002    conda-forge
xorg-libice               1.0.10               h7f98852_0    conda-forge
xorg-libsm                1.2.3             hd9c2040_1000    conda-forge
xorg-libx11               1.6.12               h36c2ea0_0    conda-forge
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxaw               1.0.14               h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxmu               1.1.3                h516909a_0    conda-forge
xorg-libxpm               3.5.13               h516909a_0    conda-forge
xorg-libxt                1.1.5             h516909a_1003    conda-forge
xorg-xextproto            7.3.0             h7f98852_1002    conda-forge
xorg-xproto               7.0.31            h7f98852_1007    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h516909a_0    conda-forge
zeromq                    4.3.4                h9c3ff4c_0    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge
zziplib                   0.13.69              h27826a3_1    conda-forge
```

</details>

diegoferigo commented:
Now few tests are failing, but the majority are passing. Definitely a step forward. I'm not sure what's missing now, I guess I need to debug it locally, from CI it's not clear to me.

diegoferigo commented:
Rebasing on latest `devel` to check if anything changes.

diegoferigo committed:
Minor edit

diegoferigo committed:
Export IGN_GAZEBO_SERVER_CONFIG_PATH from Python

diegoferigo committed:
Add missing header

diegoferigo committed:
Export all symbols in windows

diegoferigo committed:
Build plugins as MODULE in macos

diegoferigo committed:
[remove] try pining different blas backend on macos

diegoferigo committed:
Workaround zlib problem in Windows

diegoferigo committed:
Compile the Physics system with /bigobj on Windows

diegoferigo committed:
Add conda.yml


Install explicitly libode


Disable failing tests due to conda-forge's dart


[remove]

diegoferigo committed:
Adapt Ignition installation check for Windows


[remove] temporarily disable check for ign command

diegoferigo committed:
[?] import in windows with cmake_build_extension context manager

diegoferigo committed:
[remove] disable signal handling

diegoferigo committed:
[rem] run python tests verbosely

diegoferigo committed:
[move_to_py] create dot folder windows

diegoferigo commented:
> [!NOTE]  
> Cleaning up old issues and pull requests I opened over the past years that are now outdated or stale. If this is still relevant, feel free to reopen it or create a new one with updated details. Thanks!
